packages/agreement-process/src/services/readModelService.ts

```diff
@@ -318,51 +319,108 @@ async function searchTenantsByName(
 }
 
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
-function getDelegateAgreementsFilters(producerIds: TenantId[] | undefined) {
-  return producerIds && producerIds.length > 0
-    ? [
+const getDelegationLookup = (kind: DelegationKind) => [
+  {
+    $lookup: {
+      from: "delegations",
+      let: {
+        eserviceId: "$data.eserviceId",
+        consumerId: "$data.consumerId",
+        producerId: "$data.producerId",
+      },
+      pipeline: [
         {
-          $lookup: {
-            from: "delegations",
-            localField: "data.eserviceId",
-            foreignField: "data.eserviceId",
-            as: "delegations",
+          $match: {
+            $expr: {
+              $and: [
+                {
+                  $eq: ["$data.eserviceId", "$$eserviceId"],
+                },
+                {
+                  $eq: [
+                    "$data.delegatorId",
+                    kind === delegationKind.delegatedConsumer
+                      ? "$$consumerId"
+                      : "$$producerId",
+                  ],
+                },
+                {
+                  $eq: ["$data.kind", kind],
+                },
+                {
+                  $eq: ["$data.state", delegationState.active],
+                },
+              ],
+            },
           },
         },
-        {
-          $unwind: {
-            path: "$delegations",
-            preserveNullAndEmptyArrays: true,
-          },
+      ],
+      as: "delegations",
+    },
+  },
+];
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+function getProducerOrConsumerWithDelegateFilters(
+  producerIds: TenantId[],
+  consumerIds: TenantId[],
+  requesterId: TenantId
+) {
+  const requesterRole: "Consumer" | "Producer" | undefined =
+    producerIds.length === 1 && producerIds[0] === requesterId
+      ? "Producer"
+      : consumerIds.length === 1 && consumerIds[0] === requesterId
+      ? "Consumer"
+      : undefined;
```

Comment:
This is the problematic part. With this check we understand the call is coming from the frontend pages "Deleghe in erogazione" or "Deleghe in fruizione". In my opinion we should split this call in 3, `getAgreements` generic without delegation, `/consumer/agreements` without the the `consumerIds` (this is the `requesterId`), `/producer/agreements` without `producerIds` (this is the `requesterId`).

I wouldn't do it in the context of this PR but I would open a task to improve this part.

Reply:
I agree: at the moment this call is too generic and used in a hackish way by the two front-end pages. By implementing it like this, we are hard-coding a front-end hack on the backend side adding the risk of unexpected side-effects in other parts of the code where this call is used.

For example, the API Gateway [here](https://vscode.dev/github/pagopa/interop-be-monorepo/blob/develop/packages/api-gateway/src/services/agreementService.ts#L29-L30): if by chance that call is called with a single consumer filter that also corresponds to the caller organization, it will end up in the case where also delegated agreements are returned. 

I would keep the `getAgreements` as generic as possible and unrelated from any delegation logic, and implement two specific `/producer/agreements` and `/consumer/agreements` calls as you suggest.

Reply:
I'm experiencing similar issues. I also think the best way to resolve this kind of issues is to create dedicated endpoints.